src/code/sys_matrix.c

```diff
@@ -1,89 +1,1939 @@
+ *
+ * I.e. calculate \f[ A(0, 0, z) + b \f]`.
+ *
+ * @note Special case of @sa Matrix_MultVec3f with `src = { 0, 0, z }`; the same assumptions apply.
+ *
+ * @param z multiplier of unit vector in z direction.
+ * @param dest output.
+ *
+ * @remark original name: "Matrix_Position_VecZ"
+ */
+void Matrix_MultVecZ(f32 z, Vec3f* dest) {
+    MtxF* cmf = sCurrentMatrix;
+
+    dest->x = cmf->xw + cmf->xz * z;
+    dest->y = cmf->yw + cmf->yz * z;
+    dest->z = cmf->zw + cmf->zz * z;
+}
+
+/**
+ * @brief Calculates current * (src,1) and writes its x and z components to dest.
+ *
+ * The same as @sa Matrix_MultVec3f, but only applies to the x and z components; the same assumptions apply.
+ *
+ * @note Unlike the previous functions, does *not* just multiply (x, 0, z, 1) and save the x,y,z components.
+ *
+ * @param src input vector.
+ * @param dest output vector.
+ */
+void Matrix_MultVec3fXZ(Vec3f* src, Vec3f* dest) {
+    MtxF* cmf = sCurrentMatrix;
+
+    dest->x = cmf->xw + (cmf->xx * src->x + cmf->xy * src->y + cmf->xz * src->z);
+    dest->z = cmf->zw + (cmf->zx * src->x + cmf->zy * src->y + cmf->zz * src->z);
+}
+
+/**
+ * @brief Copies the matrix src into dest.
+ *
+ * @param dest matrix to copy to.
+ * @param src matrix to copy from.
+ *
+ * @remark original name: "Matrix_copy_MtxF"
+ */
+void Matrix_MtxFCopy(MtxF* dest, MtxF* src) {
+    f32 fv0;
+    f32 fv1;
+
+    // This ought to be a loop, but all attempts to match it as one have so far failed.
+    if (1) {
+        fv0 = src->mf[0][0];
+        fv1 = src->mf[0][1];
+        dest->mf[0][0] = fv0;
+        dest->mf[0][1] = fv1;
+        fv0 = src->mf[0][2];
+        fv1 = src->mf[0][3];
+        dest->mf[0][2] = fv0;
+        dest->mf[0][3] = fv1;
+    }
+    if (1) {
+        fv0 = src->mf[1][0];
+        fv1 = src->mf[1][1];
+        dest->mf[1][0] = fv0;
+        dest->mf[1][1] = fv1;
+        fv0 = src->mf[1][2];
+        fv1 = src->mf[1][3];
+        dest->mf[1][2] = fv0;
+        dest->mf[1][3] = fv1;
+    }
+    if (1) {
+        fv0 = src->mf[2][0];
+        fv1 = src->mf[2][1];
+        dest->mf[2][0] = fv0;
+        dest->mf[2][1] = fv1;
+        fv0 = src->mf[2][2];
+        fv1 = src->mf[2][3];
+        dest->mf[2][2] = fv0;
+        dest->mf[2][3] = fv1;
+    }
+    if (1) {
+        fv0 = src->mf[3][0];
+        fv1 = src->mf[3][1];
+        dest->mf[3][0] = fv0;
+        dest->mf[3][1] = fv1;
+        fv0 = src->mf[3][2];
+        fv1 = src->mf[3][3];
+        dest->mf[3][2] = fv0;
+        dest->mf[3][3] = fv1;
+    }
+}
+
+/**
+ * @brief Converts fixed-point RSP-compatible matrix to an MtxF.
+ *
+ * @param[in] src mtx to convert
+ * @param[out] dest MtxF to output to
+ *
+ * @remark original name: "Matrix_MtxtoMtxF"
+ */
+void Matrix_MtxToMtxF(Mtx* src, MtxF* dest) {
+    u16* intPart = (u16*)&src->m[0][0];
+    u16* fracPart = (u16*)&src->m[2][0];
+
+    dest->xx = ((intPart[0] << 0x10) | fracPart[0]) * (1 / (f32)0x10000);
```

Comment:
```suggestion
    dest->xx = ((intPart[0] << 16) | fracPart[0]) * (1 / (f32)0x10000);
```
To match the style of `gdSPDefMtx`

Reply:
I wonder which is clearer, actually. I'd consider changing `gdSPDefMtx` instead.

Reply:
Either is fine, I'd prefer consistency. Though also note that `Matrix_MtxFToMtx` seems to have a similar operation and also uses `0x10`